### ISSUE TITLE
Any particular reason to use whole model for ensuring "normality" of encoder outputs?

### ISSUE THREAD

**Issue by hadaev8** [ConorLazarou/AEGAN-keras#1]
Why not use Kullback-Leibler divergence? As it seems to be used in autoencoders.

**Comment by RileyLazarou** [comment]
Good question! GAN discriminators can (theoretically) learn the JS divergence between E(X) and Z for any distribution Z. So yes, it would have been simpler to use the KL divergence for Dz, but by using a second discriminator we can generalize to arbitrary distributions, which is useful if you have reason to believe a spherical normal distribution is less appropriate than some other distribution, or indeed if you don't have access to the true data generating function and are using some approximation, such as the Metropolis–Hastings algorithm. Makhzani et al. expound on this idea in their paper [Adversarial Autoencoders](https://arxiv.org/abs/1511.05644)

**Comment by hadaev8** [comment]
Huh, I tried to use kd div instead of the model and seems like I'm doing wrong something.
I used reparametrization as it used in vaes.
Like here
https://github.com/AntixK/PyTorch-VAE/blob/master/models/vanilla_vae.py#L107
But distance loss between output and original z does not work.
Have you any ideas about it?

**Comment by RileyLazarou** [comment]
No idea, sorry mate. Try posting to stack overflow with a minimal code example.
